### PR TITLE
Created init rule

### LINKED ISSUE #325: Rule: init
Disallow `nsITransferrable.init(null)`

https://github.com/mozilla/addons-validator/blob/master/docs/rules.md#entity-values

### PR DESCRIPTION
**Pull request by Nolski**
Resolves #325

### CODE REVIEW

File: src/rules/javascript/init.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { INIT } from 'messages/javascript';
+import { getNodeReference } from 'utils';
+
+export function init(context) {
+  return {
+    CallExpression: function(node) {
+      // Check if what's being called is setTimeout or setInterval
+      if (typeof node.callee.object !== 'undefined' &&
```
Comment (tofumatt) [inline src/rules/javascript/init.js]: This is pretty hard to parse. What exactly does it pass/fail for?

File: src/rules/javascript/init.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { INIT } from 'messages/javascript';
+import { getNodeReference } from 'utils';
+
+export function init(context) {
+  return {
+    CallExpression: function(node) {
+      // Check if what's being called is nsiTransferable.init()
```
Comment (Nolski) [inline src/rules/javascript/init.js]: Sorry, some of these comments were out of date.

File: src/const.js [hunk 3]
```diff
@@ -9,6 +9,7 @@ export const ESLINT_RULE_MAPPING = {
   deprecated_entities: ESLINT_WARNING,
   eval_string_arg: ESLINT_ERROR,
   global_require_arg: ESLINT_WARNING,
+  init: ESLINT_WARNING,
```
Comment (muffinresearch) [inline src/const.js]: Again this is too vague - please add some more context to the warning. It would be nice to try and infer a little bit more info from just the error code. `INIT_NULL_FIRST_ARG` or something like that.

File: src/rules/javascript/deprecated_entities.js [hunk 4]
```diff
@@ -27,11 +27,11 @@ export function deprecated_entities(context) {
           node.callee.property.type === 'Identifier' &&
           node.callee.object.type === 'Identifier') {
 
-        let objectName = getNodeReferenceName(context, node.callee.object);
+        let object = getNodeReference(context, node.callee.object);
```
Comment (muffinresearch) [inline src/rules/javascript/deprecated_entities.js]: Something like this it's probably nice to add something to the var name so we know what object we're talking about later on in the code.

File: tests/rules/javascript/test.init.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import JavaScriptScanner from 'scanners/javascript';
+import { VALIDATION_WARNING } from 'const';
+import * as messages from 'messages';
+
+describe('init', () => {
+  it('should not allow init called with null', () => {
```
Comment (muffinresearch) [inline tests/rules/javascript/test.init.js]: Does this rule handle aliasing e.g:

```javascript
var foo = nsITransferable.init;
foo(null);
```

and

```javascript
var foo = nsITransferable;
foo.init(null);
```

File: tests/rules/javascript/test.init_null_arg.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    var jsScanner = new JavaScriptScanner(code, 'badcode.js');
+
+    return jsScanner.scan()
+      .then((validationMessages) => {
+        assert.equal(validationMessages.length, 0);
+      });
+  });
+
+  it('should not allow init called with an aliased varaible and null', () => {
+    var code = 'var foo = nsITransferable; foo.init(null);';
```
Comment (muffinresearch) [inline tests/rules/javascript/test.init_null_arg.js]: Nice, can you add one more test to show it catches this?
```
var foo = nsITransferable.init;
foo(null);
```

**Comment by muffinresearch** [comment]
r+wc
